pio/7_segment/pio_7_segment_library/7_segment_lib.pio

```diff
@@ -0,0 +1,50 @@
+;;
+;; Copyright (c) 2023 mjcross
+;;
+;; SPDX-License-Identifier: BSD-3-Clause
+;;
+
+.program seven_segment
+
+;; SIDE pins:   digit enable lines
+;; OUT pins:    segments (multiplexed)
+
+.side_set 4                             ; leaves one `delay` bit
+
+                                        ; note: for the digits to look the same brightness it's
+                                        ; important to show them for the same amount of time
+.wrap_target
+    pull noblock    side 0b0001         ; TX FIFO (or X) -> OSR     keep showing 1000's for 1 tick
+                                        ; if the FIFO is empty then the OSR is loaded from X
+public entry_point:
+    mov x, osr      side 0b0000         ; save the segments in X    and show no digits for 1 tick
+                                        
+    out pins, 8     side 0b1000  [1]    ; OSR[07..00] -> segments   and show 1's digit for 2 ticks
+    out pins, 8     side 0b0100  [1]    ; OSR[15..08] -> segments   and show 10's digit for 2 ticks
+    out pins, 8     side 0b0010  [1]    ; OSR[23..16] -> segments   and show 100's digit for 2 ticks
+    out pins, 8     side 0b0001         ; OSR[31..24] -> segments   and show 1000's digit for 1 tick
+.wrap
+;; 6 instructions
+
+
+% c-sdk {
+static inline void seven_segment_sm_init (PIO pio, uint sm, uint offset, 
+                                          uint segment_pinbase, uint digit_pinbase) {
+    // create new sm config
+    pio_sm_config config = seven_segment_program_get_default_config (offset);
+
+    // configure the common cathodes and segment pin groups
+    sm_config_set_out_pins (&config, segment_pinbase, 8);
+    sm_config_set_sideset_pins (&config, digit_pinbase);
+
+    // configure the clock divider
+    float div = clock_get_hz (clk_sys) * 10e-3;
```

Comment:
I may be missing something, but it seems that if clk_sys is 125MHz (for instance) then div will become 125000. This is higher than 65536, which is the maximum supported by the clock divider. I think this needs a check and if the clock is too high, the divisor would just be set to 65536. This will produce a refresh rate of around 2kHz on the display, which should still be ok.

Reply:
Good shout - fixed with https://github.com/mjcross/pico-examples/commit/25ac8f45374d1bacec096dae20ec09623164a2f7